### PR TITLE
DynamicUser=1 state directory back migration

### PR DESCRIPTION
**Pull request by poettering**
A fix for #12131

### CODE REVIEW

File: src/core/execute.c [hunk 1]
```diff
@@ -2171,6 +2171,38 @@ static int setup_exec_directory(
                                 goto fail;
 
                 } else {
+                        _cleanup_free_ char *target = NULL;
+
+                        if (type != EXEC_DIRECTORY_CONFIGURATION &&
```
Comment (yuwata) [inline src/core/execute.c:2176]: Please add comments why runtime directories are not excluded here.

File: src/core/execute.c [hunk 2]
```diff
@@ -2171,6 +2171,38 @@ static int setup_exec_directory(
...
+
+                                        /* Hmm, apparently DynamicUser= was once turned on for this service,
+                                         * but is no longer. Let's move the directory back up. */
+
+                                        if (unlink(p) < 0) {
+                                                r = -errno;
+                                                goto fail;
+                                        }
+
+                                        if (rename(q, p) < 0) {
```
Comment (keszybz) [inline src/core/execute.c:2199]: I think it's be worthwhile to output a log_info() here that the directory is being renamed. This should happen very rarely and can seem quite magical if someone doesn't know about the private/ directory.

For symmetry, a message should also be emitted when moving *to* private/.

**Comment by poettering** [comment]
merging due to green label, will prepare follow-up right after.

**Comment by poettering** [comment]
Follow-up in #12877

**Pull request merged**